### PR TITLE
Gateway cascade enhancements (and some more)

### PR DESCRIPTION
**Pull request by ebaauw**
Several enhancements for gateway cascade functionality, see #287:
- Cascade setting `action.on` and `action.bri`, with or without `action.transitiontime` from the REST API;
- Cascade recalling scenes to the Hue bridge: support Hue bridge syntax for scene commands, see #29;
- Cascade all commands sent by the Hue dimmer switch in group mode;
- Cascade all commands sent by the IKEA Trådfri remote in group mode.

Notes:
- The Hue dimmer switch actually sends multiple _Step_ commands (~one every second) while holding the _DimUp_ or _DimDown_ button; the Trådfri remote sends a single _Move_ command and relies on the _Stop_ command (on long release) to stop the current movement;
- The Trådfri remote sends a _Toggle_ command on the _On/Off_ button.   As this is not supported by the REST API, `action.on` is set to true (false) when local group `state.all_on` is false (true);
- Currently, the _DimUp_ and _DimDown_ buttons (for both the Hue and the Trådfri) can only be cascaded to a Hue bridge, as deCONZ hasn't yet implemented `bri_inc` (see #84).

Todo:
- Cascade setting `action.ct` and `action.xy` from the REST API.

Other changes:
- Change ubisys dimmer buttonevents;
- Only expose ZHASwitch sensor resources for _Analog Input_ and _Multistate Input_ clusters for the Xiaomi Smart Cube, not for other devices.  Apparently, sometimes other Xiaomi devices advertise these clusters.

### CODE REVIEW

File: de_web_plugin.cpp [hunk 1]
```diff
@@ -2694,13 +2694,17 @@ void DeRestPluginPrivate::addSensorNode(const deCONZ::Node *node)
 
                 case ANALOG_INPUT_CLUSTER_ID:
                 {
-                    fpSwitch.inClusters.push_back(ci->id());
+                    if (modelId == QLatin1String("lumi.sensor_cube")) {
```
Comment (manup) [inline de_web_plugin.cpp]: We may need to extend this whitelist in future, I'm not sure but I think there were other sensors using the analog/multi-state clusters too.

**Comment by manup** [comment]
Thanks a lot, I'll merge the PR now but can't test it currently (back in end of January).

Maybe in future the support for proper external web hooks in rules replaces the cascades with a simpler general way to forward REST API requests to an arbitrary destination.

See also https://github.com/dresden-elektronik/deconz-rest-plugin/issues/347

Luckily the implementation will be fairly simple due QNetworkManager. :)

File: de_web_plugin.cpp (hunk 1 above)
Comment (ebaauw) [reply de_web_plugin.cpp]: I double-checked that I introduced these clusters when adding support for the Smart Cube.  Also, the handling of the attribute values is specific to the cube.

Comment (ebaauw) [reply de_web_plugin.cpp]: Alternatively we could to blacklist the “funny” sensors (so far I’ve seen reports of lumi.sensor_ht and lumi.switch_86sw2 showing additional endpoints with these clusters).

**Comment by ebaauw** [comment]
> Maybe in future the support for proper external web hooks in rules replaces the cascades with a simpler general way to forward REST API requests to an arbitrary destination.

Maybe, maybe not.  There’s something to say for webhooks, especially if they can be attached to rule actions.  You’d be able to interact with any REST api, not just the deCONZ/Hue API.  With cascade, you actually forward ZigBee commands as deCONZ/Hue REST API calls.  I’m amazed how fast this works, when the commands are sent by ZigBee switches or sensors, compared to commands sent by the gateway (looks like the gateway only cascades these when it reads back the groupcasts).

The translation of ZigBee commands to REST API calls is somewhat crude.  I’d love someday to design a more consistent and complete API, with different endpoints for (write-only) commands, `/actions/...`, read-only attributes, `/state/...`, and writeable attributes, `/config/...`.

**Comment by manup** [comment]
Agree, they both have their use cases. Setup for cascades is also easier since it's just a group a forward to group b mapping.

**Pull request merged**